processor/cumulativetodeltaprocessor/processor.go

```diff
@@ -16,35 +16,99 @@ package cumulativetodeltaprocessor
 )
 
 type cumulativeToDeltaProcessor struct {
-	metrics []string
-	logger  *zap.Logger
+	metrics         []string
+	logger          *zap.Logger
+	deltaCalculator awsmetrics.MetricCalculator
 }
 
 func newCumulativeToDeltaProcessor(config *Config, logger *zap.Logger) *cumulativeToDeltaProcessor {
 	return &cumulativeToDeltaProcessor{
-		metrics: config.Metrics,
-		logger:  logger,
+		metrics:         config.Metrics,
+		logger:          logger,
+		deltaCalculator: newDeltaCalculator(),
 	}
 }
 
 // Start is invoked during service startup.
-func (mgp *cumulativeToDeltaProcessor) Start(context.Context, component.Host) error {
+func (ctdp *cumulativeToDeltaProcessor) Start(context.Context, component.Host) error {
 	return nil
 }
 
 // processMetrics implements the ProcessMetricsFunc type.
-func (mgp *cumulativeToDeltaProcessor) processMetrics(_ context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+func (ctdp *cumulativeToDeltaProcessor) processMetrics(_ context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+	inputMetricSet := make(map[string]bool)
+	for _, name := range ctdp.metrics {
+		inputMetricSet[name] = true
+	}
+
+	resourceMetricsSlice := md.ResourceMetrics()
+
+	for i := 0; i < resourceMetricsSlice.Len(); i++ {
+		rm := resourceMetricsSlice.At(i)
+		ilms := rm.InstrumentationLibraryMetrics()
+		for i := 0; i < ilms.Len(); i++ {
+			ilm := ilms.At(i)
+			metricSlice := ilm.Metrics()
+			for j := 0; j < metricSlice.Len(); j++ {
+				metric := metricSlice.At(j)
+				_, ok := inputMetricSet[metric.Name()]
+				if ok {
+					if metric.DataType() == pdata.MetricDataTypeSum {
+						dataPoints := metric.Sum().DataPoints()
+
+						for i := 0; i < dataPoints.Len(); i++ {
+							fromDataPoint := dataPoints.At(i)
+							labelMap := make(map[string]string)
+
+							fromDataPoint.LabelsMap().Range(func(k string, v string) bool {
+								labelMap[k] = v
+								return true
+							})
+
+							result, _ := ctdp.deltaCalculator.Calculate(metric.Name(), labelMap, fromDataPoint.Value(), fromDataPoint.Timestamp().AsTime())
+
+							fromDataPoint.SetValue(result.(delta).value)
+							fromDataPoint.SetStartTimestamp(pdata.TimestampFromTime(result.(delta).prevTimestamp))
+						}
+						metric.Sum().SetAggregationTemporality(pdata.AggregationTemporalityDelta)
+					}
+				}
+			}
+		}
+	}
 	return md, nil
 }
 
 // Shutdown is invoked during service shutdown.
-func (mgp *cumulativeToDeltaProcessor) Shutdown(context.Context) error {
+func (ctdp *cumulativeToDeltaProcessor) Shutdown(context.Context) error {
 	return nil
 }
+
+func newDeltaCalculator() awsmetrics.MetricCalculator {
+	return awsmetrics.NewMetricCalculator(func(prev *awsmetrics.MetricValue, val interface{}, timestamp time.Time) (interface{}, bool) {
+		result := delta{value: val.(float64), prevTimestamp: timestamp}
+
+		if prev != nil {
+			deltaValue := val.(float64) - prev.RawValue.(float64)
```

Comment:
I had to look through `awsmetrics` to see what the calculator does here. This looks good to me, although you might consider for monotonic cumulatives to insert a "Reset heuristic" the way Prometheus has. The current logic looks like it will deliver a negative delta when a cumulative value resets, and you might want to interpret that as a zero depending on your knowledge.

See https://github.com/lightstep/opentelemetry-prometheus-sidecar/blob/main/retrieval/series_cache.go#L331 for a similar handler used to convert Prometheus cumulatives w/o a start time to cumulatives with a start time. If you pass through similar logic first, you won't see the negative delta (for monotonic cumulatives).

Approving this PR because the implied logic fix probably belongs in `awsmetrics` (?).

Reply:
Thanks Josh. Will check and update here.

Reply:
I am also curious about this. It looks like even if the aws library is updated, the logic above (`if prev != nil`) means that if a producer sends only 1 value of a cumulative metric, the delta will never be reported. Is that correct?

Reply:
@jmacd in order to determine a reset we should initially use "startTime" and if that is missing (0), fallback to the prometheus way to detect reset?